### ISSUE TITLE
Assertion failure calling GridFunction.GetGradient() - incorrect dimensions

### ISSUE THREAD

**Issue by jayjaybillings** [mfem/mfem#405]
I get the following error when calling GridFunction.GetGradient():
```
Assertion failed: (inva.Height() == a.Width()) is false: incorrect dimensions
 ... in function: void mfem::CalcInverse(const mfem::DenseMatrix&, mfem::DenseMatrix&)
 ... at line 3159 of file: /home/bkj/programs/mfem-dev-20171119/linalg/densemat.cpp
```
My debugger reports the following in this frame:
```
(gdb) f 3
#3  0x0000000000444cc5 in mfem::CalcInverse (a=..., inva=...)
    at /home/bkj/programs/mfem-dev-20171119/linalg/densemat.cpp:3159
3159	   MFEM_ASSERT(inva.Height() == a.Width(), "incorrect dimensions");
(gdb) p inva
$3 = (mfem::DenseMatrix &) @0x7fffffffbb00: {<mfem::Matrix> = {<mfem::Operator> = {
      _vptr.Operator = 0x59eee0 <vtable for mfem::DenseMatrix+16>, height = 2, 
      width = 2}, <No data fields>}, data = 0x828760, capacity = 4}
(gdb) p a
$4 = (const mfem::DenseMatrix &) @0x7fffffffd260: {<mfem::Matrix> = {<mfem::Operator> = {
      _vptr.Operator = 0x59eee0 <vtable for mfem::DenseMatrix+16>, height = 2, 
      width = 1}, <No data fields>}, data = 0x8286e0, capacity = 2}
(gdb) 
```
I am writing a heat equation solver and trying to compute a heat conduction boundary term, 
```
q = -k*grad(u)
```
where k is the conductivity and grad(u) is the gradient of the temperature. The weak form of this term is (q\dot n, v). I configure the boundary condition as a BoundaryNormalLFIntegrator with a subclass of VectorCoefficient as follows:
```cpp
/**
 * This coefficient computes the heat conduction flux, q = -k*grad(T) on the
 * boundary.
 */
class ConductionFluxCoefficient : public mfem::VectorCoefficient {

	mfem::GridFunction & u;

	double k;

public:
	ConductionFluxCoefficient(int _dim, mfem::GridFunction _u) :
		mfem::VectorCoefficient::VectorCoefficient(_dim), u(_u), k(1.0) {

	}

	virtual ~ConductionFluxCoefficient() {};

	virtual void setConductivity(double _k) {
		k = _k;
	}

	virtual void Eval(mfem::Vector & v, mfem::ElementTransformation & T,
			const mfem::IntegrationPoint & ip)  {
	    // Set the size of the vector to match the problem dimensionality
	    v.SetSize(vdim);
	    // v = grad(u)
	    u.GetGradient(T,v);
	    // v = k*v = k*grad(u)
	    v *= -k;
	}
};

//... Later, in my TimeDependentOperator:
conductionCoeff.setConductivity(conductivity);
forcingVector.AddBoundaryIntegrator(
	new BoundaryNormalLFIntegrator(conductionCoeff));
```
My toy problem is a simple 2D plate with a temperature on the entire boundary of 1000K. I'm very new to mfem and still learning the API, so I figure I've called it incorrectly. Any suggestions would be appreciated!

**Comment by tzanio** [comment]
Hi @jayjaybillings,

Can you clarify the weak formulation and boundary conditions for the equation you'd like to solve?

For example, the weak formulation of `-div (k grad(u)) = f` does involve the boundary term you mentioned, but usually that's either zero due to Dirichlet (`v = g`) or pure Neumann (`q.n = 0`) boundary conditions, or is replaced by a linear combination of the boundary integral `(u,v)` on the left-hand side and the boundary integral `(g,v)` on the right-hand side due to Robin (`q.n = a u + b g`) boundary conditions. 

If I understand correctly, you are trying to impose non-zero Dirichlet conditions on `u` on the whole boundary? That does not require extra boundary terms and can be handled by elimination in the assembled linear system. See Example 1 for the static and Example 16 for the time-dependent case.

Hope this helps,
Tzanio

**Comment by jayjaybillings** [comment]
Thanks for the response @tzanio! You really drilled into the source of my problems, I think.

My original strategy was to solve it as you suggested with T=1000K on the boundary and to eliminate the dofs, but that hasn't worked for me. My new strategy was to try use a flux boundary condition that uses Fourier's Law. Then,
```
q = -kgrad(u)
```
In this case the full weak form is, (I think):
```
(λu,v) - (λ∇u,∇v) - (q \dot n, v) = 0
```
where n is the normal vector at the surface. Adding that last term as a BoundaryNormalLFIntegrator is what caused the problem I originally reported.

But, I should probably start at the beginning. I am trying to solve the fully time dependent heat equation:
```
(\rho c)du/dt - div (k grad(u)) = f
```
with, for now, ```f=0``` and ```a = k/(\rho c)``` such that
```
du/dt = div (a grad(u))
```
I have been looking at several of the examples and the Volta and Joule miniapps as part of this. The solution to this is, like in ex16,
```
du/dt = (M^-1)(Ku)
```
which, for now, I am trying to solve using RK4 so that I can check the intermediate results easily:
```cpp
void ThermalOperator::Mult(const Vector &u, Vector &dudt) const {
     stiffnessMatrix->Mult(u,z); // z = Ku
     z.Neg() // z = -Ku. Needed since the DiffusionIntegrator docs says there is a minus in front of it.
     mSolver.Mult(z,dudt); // dudt = (M^-1)z
}
```
This is all on an H^1 nodal finite element space with a tet mesh, but I'm thinking of switching it to a simple quad mesh to write a 2D "patch test." I set this up using the following code:
```cpp
        // Create the initial conditions, T=300
	ConstantCoefficient temp0(initialInteriorTemperature);
	temperature.ProjectCoefficient(temp0);

        // Project boundary values T=1000K. "dbc" is a simple DirichletBoundaryCoundition
        // wrapper that just holds the ConstantCoefficient, boundary attribute array, and list
        // list of essential degrees of freedom.
	temperature.ProjectBdrCoefficient(dbc.getCoefficient(),
			dbc.getBoundaryAttributes());

	// Assemble the forcing vector - no boundary or domain integrators since f = 0.
	forcingVector.Assemble();

	// Create the mass matrix as a bilinear form with the mass integrator.
	massMatrix = make_unique<BilinearForm>(&feSpace);
	massMatrix->AddDomainIntegrator(new MassIntegrator);
	massMatrix->Assemble(skip_zeros);
	massMatrix->EliminateEssentialBC(dbc.getBoundaryAttributes(),temperature,
			forcingVector);
	massMatrix->Finalize(skip_zeros);

	// Configure the mass matrix solver.
	mSolver.SetPreconditioner(mPreconditioner);
	mSolver.SetOperator(massMatrix->SpMat());
	mSolver.iterative_mode = false;
	mSolver.SetRelTol(1.0e-9);
	mSolver.SetAbsTol(0.0);
	mSolver.SetMaxIter(100);
	mSolver.SetPrintLevel(0);

	// Setup the stiffness matrix using the diffusion kernel
	ConstantCoefficient alphaCoeff(alpha);
	stiffnessMatrix = make_unique<BilinearForm>(&feSpace);
	stiffnessMatrix->AddDomainIntegrator(new DiffusionIntegrator(alphaCoeff));
	stiffnessMatrix->Assemble(skip_zeros);
	stiffnessMatrix->EliminateEssentialBC(dbc.getBoundaryAttributes(),temperature,
			forcingVector);
	stiffnessMatrix->Finalize(skip_zeros);
```
After each time step, I reimpose the boundary conditions with
```cpp
	temperature.ProjectBdrCoefficient(dbc.getCoefficient(),
			dbc.getBoundaryAttributes());
```
What I am seeing is that the temperature always decreases in the interior of the plate. Even if I remove the z.Neg() in the ThermalOperator::Mult() operation, the temperature always decreases. I have used coarse and fine meshes in 2D & 3D, but I get the behavior each time. Over a 0.2 second time frame, the interior temperature will decrease to about 240K, even though the boundary remains fixed at 1000K. The lowest temperature is on the elements immediately adjacent to the boundary. I figured that I've had bad luck with a bug, not good luck by discovering perpetual motion. ;-)

Thanks again for getting back to me. I'm confident that this is just a usage problem since I'm new to MFEM (I absolutely love it btw!). Any suggestions you have would be very helpful!

**Comment by juan42** [comment]
So the initial condition is what you expect?

I found it way more convenient using `FormLinearSystem` in order to apply boundary conditions correctly. Maybe you find something useful in an [older issue](https://github.com/mfem/mfem/issues/142). I handle time dependent Dirichlet there.

**Comment by Dan2997925** [comment]
Hi @jayjaybillings ,

I have not looked at your code or tried to run it. But from reading the comments above, I suspect the problem is with the time integration, not with the original Dirichlet boundary conditions. In the Joule miniapp I solve thermal diffusion implicitly, with a time step that is *larger* than the physical diffusion time (the time it takes for the temperature to diffuse across the element). I know it is counter-intuitive that larger time steps are better, but this is my observation, and I think there is theory that backs this up.

PS as an experiment, if your grid is Cartesian, try mass-lumping on the mass matrix to make it diagonal. Explict RK should work for this special case.

**Comment by jayjaybillings** [comment]
@juan42 Thanks for your response! I expect that the interior of the plate will heat up with time since the temperature of 1000K is on the boundary. The temperature of the interior nodes may not equal 1000K in the future, but it should increase since the heat has to flow from the boundary (1000K) to the interior (300K) according to thermodynamics.

In #142 you say that you are reinitializing b with every time step. Was that just a simple one-line fix like this?
```cpp
  std::cout << "Forming linear system with size: " << K_.Height() << std::endl;
  b_ = Vector(); // Re-initializes b_
  div_grad_->FormLinearSystem(ess_tdof_list_, *u_, *rhs_, K_, X_, b_, true);
}
```

**Comment by jayjaybillings** [comment]
@Dan2997925 Thanks for the response! Yes, the theory backs that up. I just figured the answer would at least increase for small explicit time steps. I'll give it a try and see what happens.

Thanks for the mass-lumping tip too.

**Comment by juan42** [comment]
Since `FormLinearSystem` overwrites every entry in b_ that should suffice, yes.

In your case it looks like ThermalOperator::Mult does not have the correct boundary conditions in `dudt` after the solve or in `u` before the solve.

**Comment by juan42** [comment]
@jayjaybillings have you sorted out the issue? If not and you are still interested in a heat radiation "miniapp" i can provide a working one.

**Comment by jayjaybillings** [comment]
Thanks @juan42 and @Dan2997925 - I was able to fix the issue based on your ideas. (@juan42 I would still be interested in seeing your miniapp if you are willing to share.)

It required doing both of the things you guys suggested to get the behavior I expect. Apologies for the delay getting back to you - I've been on other projects. Here's what I had to do:
1. I switched to Form{System/Linear}Matrix instead of EliminateEssentialBCs. The former seemed to treat the boundary better and provided a less incorrect result.
2. I fixed a bug in how I was handling boundary condition enforcement in the solver.
3. I  added a recoverSolution operation as in @juan42's example in #142. This was something I thought I needed to do anyway, so I was really glad to see it in #142.
4. Switched to an implicit solver with a large time step.

After fixing 1-3, the solver behaved mostly as I would expect during a explicit integration and because of stiffness issues would still drive some parts of the mesh off course. As @Dan2997925 suggested, switching to an implicit solve and doing a time step on the order of the diffusion time scale for the material worked. So for a 2D element with side length ```L```, I computed the time step as follows
```
L = 2.0*sqrt(\alpha t)
t = L^2/(4.0*\alpha)
```
With this time step, the heat flow obeys the second law and the results look quite reasonable in both 2D and 3D. 

Today I'm working to setup a validation/patch test to check the actual correctness of the solution. Thanks very much for your help guys!
